apps/io_tester/io_tester.cc

```diff
@@ -657,6 +665,99 @@ class write_io_class_data : public io_class_data {
+        }
+
+        const auto fname = get_filename(_file_id_to_remove);
+        ++_file_id_to_remove;
+
+        return remove_file(fname).then([]{
+            return make_ready_future<size_t>(1u);
+        });
+    }
+
+    void emit_results(YAML::Emitter& out) override {
+        out << YAML::Key << "average latency" << YAML::Value << average_latency();
+    }
+
+private:
+    uint64_t files_count() const {
+        return *_config.files_count;
+    }
+
+    bool all_files_removed() const {
+        return files_count() <= _file_id_to_remove;
+    }
+
+    sstring get_filename(uint64_t file_id) const {
+        return format("{}/test-{}-shard-{:d}-file-{}", _dir_path, name(), this_shard_id(), file_id);
+    }
+
+    future<> do_start_on_directory(sstring path) {
+        _dir_path = std::move(path);
+
+        // When we have many files it is easy to exceed the limit of open file descriptors.
+        // To avoid that the limit is divided between shards (leaving some room for other jobs).
+        const uint64_t max_concurrency = static_cast<uint64_t>((1024u / smp::count) * 0.8);
+        return max_concurrent_for_each(boost::irange(UINT64_C(0), files_count()), max_concurrency, [this] (uint64_t file_id) {
+            const auto fname = get_filename(file_id);
+            const auto fsize = _config.file_size / files_count();
+            const auto flags = open_flags::rw | open_flags::create;
+
+            file_open_options options;
+            options.extent_allocation_size_hint = fsize;
+            options.append_is_unlikely = true;
+
+            return open_file_dma(fname, flags, options).then([fsize] (auto f) mutable {
+                return do_with(std::move(f), [fsize] (auto f) {
+                    return f.size().then([f, fsize] (uint64_t pre_truncate_size) mutable {
+                        return f.truncate(fsize).then([f, fsize, pre_truncate_size] () {
+                            if (pre_truncate_size >= fsize) {
+                                return make_ready_future<>();
+                            }
+
+                            const uint64_t buffer_size{256ul << 10};
+                            const uint64_t buffers_count{static_cast<uint64_t>(fsize / buffer_size) + 1u};
+
+                            return do_with(boost::irange(UINT64_C(0), buffers_count), [f, buffer_size] (auto& buffers_range) mutable {
+                                return max_concurrent_for_each(buffers_range.begin(), buffers_range.end(), 64, [f, buffer_size] (auto buffer_id) mutable {
+                                    auto source_buffer = allocate_and_fill_buffer(buffer_size);
+                                    auto write_position = buffer_id * buffer_size;
+
+                                    return f.dma_write(write_position, source_buffer.get(), buffer_size).discard_result();
+                                });
+                            }).then([f]() mutable {
+                                return f.flush();
```

Comment:
File should also be close()d

Reply:
Okay, a call to `close()` has been added. It will be uploaded in the next patch-set.

I did not add it previously, because I thought that the file is closed when the last object holding the reference to FD is destroyed. In this particular case I thought that the destruction of an object that is held in `do_with()` closes the file.

From `seastar/src/core/file.cc`:

```
posix_file_handle_impl::~posix_file_handle_impl() {
    if (_refcount && _refcount->fetch_add(-1, std::memory_order_relaxed) == 1) {
        ::close(_fd);
        delete _refcount;
    }
}
```

Reply:
Yes, it does, but explicit is better than implicit